### PR TITLE
Improve python3 support of test_wpt

### PR DESCRIPTION
**Pull request by svillar**
We are not there yet even with these changes, that's why I kept the xfail around. There are several issues with manifest file writting that need to be addressed first. However with these changes, tests pass in python3 as long as we don't update de manifest.

### CODE REVIEW

File: tools/manifest/utils.py [hunk 1]
```diff
@@ -57,14 +57,14 @@ def to_os_path(path):
 def git(path):
     # type: (bytes) -> Optional[Callable[..., bytes]]
     def gitfunc(cmd, *args):
-        # type: (bytes, *bytes) -> bytes
+        # type: (bytes, *bytes) -> AnyStr
```
Comment (jgraham) [inline tools/manifest/utils.py]: Doesn't this always return `Text` now?

File: tools/manifest/sourcefile.py [hunk 2]
```diff
@@ -242,9 +242,7 @@ def __getstate__(self):
 
         if "__cached_properties__" in rv:
             cached_properties = rv["__cached_properties__"]
-            for key in rv.keys():
-                if key in cached_properties:
-                    del rv[key]
+            rv = {key:value for key, value in iteritems(rv) if key not in cached_properties}
```
Comment (jgraham) [inline tools/manifest/sourcefile.py]: Why this change?

**Review by jgraham — CHANGES_REQUESTED** [review]

File: tools/manifest/utils.py (hunk 1 above)
Comment (svillar) [reply tools/manifest/utils.py]: ACK

File: tools/manifest/sourcefile.py (hunk 2 above)
Comment (svillar) [reply tools/manifest/sourcefile.py]: Seems unrelated indeed, but python3 complains because we're indeed changing a dictionary while iterating over it, that's why I decided to use comprehension here.

Comment (jgraham) [reply tools/manifest/sourcefile.py]: Oh right, yes.

I wonder if this is performance-critical enough to write the whole function as:

```
if "__cached_properties__" in self.__dict__:
    cached_properties = self.__dict__["__cached_properties__"]
    rv = {key:value for key, value in iteritems(self.__dict__) if key not in cached_properties}
else:
    rv = self.dict.copy()
return rv
```
Maybe not, since we just spend so much time parsing HTML that a dict copy doesn't matter. But perhaps it would be worth checking for a performance regression when we rebuild the manifest with this branch?

**Review by jgraham — APPROVED** [review]

File: tools/manifest/sourcefile.py (hunk 2 above)
Comment (svillar) [reply tools/manifest/sourcefile.py]: I'll get some numbers

File: tools/manifest/sourcefile.py [hunk 3]
```diff
@@ -304,7 +302,7 @@ def url(self):
 
     @cached_property
     def hash(self):
-        # type: () -> bytes
+        # type: () -> Text
```
Comment (jgraham) [inline tools/manifest/sourcefile.py]: I kind of feel like this ought to be bytes, but I don't know what the impact of changing the `encode` in Py3 to a `decode` in Py2 is.

File: tools/wpt/testfiles.py [hunk 4]
```diff
@@ -299,7 +299,7 @@ def affected_by_wdspec(test):
 
     def affected_by_interfaces(file_contents):
         # type: (Union[bytes, Text]) -> bool
-        if len(interfaces_changed_names) > 0:
+        if interfaces_changed_names:
```
Comment (jgraham) [inline tools/wpt/testfiles.py]: This seems wrong; in Python 3:
```
In [4]: bool(map(lambda x: x, []))
Out[4]: True
```

i.e. this will evaluate as True even if the `interfaces_changed_names` map is empty, because the iterator will always evaluate as True. I think you could do `if interfaces_changed` instead?

**Review by jgraham — CHANGES_REQUESTED** [review]

File: tools/wpt/testfiles.py (hunk 4 above)
Comment (svillar) [reply tools/wpt/testfiles.py]: Good point. I'll change it

File: tools/manifest/sourcefile.py (hunk 3 above)
Comment (svillar) [reply tools/manifest/sourcefile.py]: Right, I agree with you. I'll try to look for a different solution.

File: tools/manifest/utils.py [hunk 5]
```diff
@@ -55,16 +55,16 @@ def to_os_path(path):
...
 
 def git(path):
-    # type: (bytes) -> Optional[Callable[..., Text]]
+    # type: (bytes) -> Optional[Callable[..., bytes]]
     def gitfunc(cmd, *args):
-        # type: (bytes, *bytes) -> Text
+        # type: (bytes, *bytes) -> bytes
         full_cmd = ["git", cmd] + list(args)
         try:
-            return subprocess.check_output(full_cmd, cwd=path, stderr=subprocess.STDOUT).decode('utf8')
```
Comment (jgraham) [inline tools/manifest/utils.py]: So I thought this change was correct (or at least aligns with our general strategy of assuming process output is UTF8); are you still planning to land it or is it in a different PR?

Comment (svillar) [reply tools/manifest/utils.py]: I am not sure if we should do that for git commands output. There are actually a lot of ramifications as you have seen (like having binary hashes for example) of switching to a Text output in this case.

I agree that the general strategy is to assume UTF8 output for process calls but for this particular case I think we can make an exception. Do you prefer instead to keep the decode() and then encode() as needed?

Comment (jgraham) [reply tools/manifest/utils.py]: I see. I think if we are taking the position that command output is text then we should be consistent with that. If we want to then represent hashes as the actual byte value rather than text we should convert down. But thinking about it they end up in the manifest JSON, so we have to represent them as text. So I think my earlier comment was wrong, sorry.

Comment (svillar) [reply tools/manifest/utils.py]: So... can we land it as is?

Comment (jgraham) [reply tools/manifest/utils.py]: No, I think we should reapply the changes that you reverted so that the git output is converted to a string, and the hash is a string.

File: tools/manifest/sourcefile.py (hunk 2 above)
Comment (jgraham) [reply tools/manifest/sourcefile.py]: Did you try measuring this? I don't expect it to be a big problem, but I'm at least interested.

Comment (svillar) [reply tools/manifest/sourcefile.py]: Oh I forgot to reply to this. Yes, I did take measurements. I got the following after running the `test_wpt.py` tests before & after. 

average duration (in seconds)
before: 1,98 e-06  
after:    5,61 e-06

There are indeed differences but they're so small that I think we can live with that. As an example I got 74k meassurements running those tests. If I sum all the durations I get 0,4s after and 0,15s before, so less than 0,3s of difference running all the `test_wpt.py` tests.

Comment (jgraham) [reply tools/manifest/sourcefile.py]: Interesting. I'm happy for this to land now, but maybe file an issue (or just make a followup PR) to take that change. Saving a third of a second on a full rebuild isn't much, but it is within the realm of small improvements that are worthwhile when you're trying to optimise.

**Review by jgraham — APPROVED** [review]

File: tools/manifest/sourcefile.py (hunk 2 above)
Comment (svillar) [reply tools/manifest/sourcefile.py]: OK, I'll do
